src/signature.rs

```diff
@@ -116,15 +158,26 @@ pub struct Decoded {
 
 impl Decoded {
     /// Checks the properties of individual integers in a DER signature.
-    fn is_valid_integer(int_bytes: &[u8]) -> bool {
+    fn is_valid_integer(int_bytes: &[u8]) -> Result<(), InvalidDerInteger> {
         match int_bytes {
-            // Zero-length integers are not allowed.
-            [] => false,
+            [] => Err(InvalidDerInteger::ZeroLength),
             // Null bytes at the start are not allowed, unless it would otherwise be interpreted as
             // a negative number.
-            [0x00, next, ..] => next & 0x80 != 0,
+            [0x00, next, ..] => {
+                if next & 0x80 != 0 {
+                    Ok(())
+                } else {
+                    Err(InvalidDerInteger::LeadingNullByte)
+                }
```

Comment:
This could be
```rust
(next & 0x80 != 0).then_some(()).ok_or(InvalidDerInteger::LeadingNullByte)
```
I'm not sure whether that's better. :woman_shrugging: